File: interpreter.c
#include <stdio.h>
int main(int argc, char** argv)
{
  char input[4096] = {0};
  char tape [30000] = { 0 };
  char* ptr = tape;
  puts("BF Interpreter 1.0");
  puts("Press Ctrl + c to Exit\n");

  while(1)
  {
    fputs("BF> ", stdout);
    fgets(input, 4096, stdin);
    int track = 0;

    for(int i = 0; i<4096; i++)
    {
      //Check for well formed input
      if(input[i] == '[')
      {
        track++;
      }
      else if(input[i] == ']')
      {
        track--;
      }
    }

    //check to see how many missing brackets
    if(track<0)
    {
      fputs("You are missing a [\n", stdout);
      continue;
    }
    else if(track>0)
    {
      fputs("Your are missing a ]\n", stdout);
      continue;
    }

    
  }
  return 0;
}

File: src/bf_compiler/Interpreter.java
package bf_compiler;

import java.util.Scanner;

public class Interpreter 
{

	public static void main(String[] args) 
	{
		Scanner input = new Scanner(System.in);
		System.out.println("BF Intterpreter 1.0");
		System.out.println("Press Ctrl + C to exit");
		int tape[] = new int[30000];
		int pointer = 0;
		while (true)
		{
			System.out.print("BF>");
			String command = input.nextLine();
			int tracker = 0;
			for(int i = 0; i<command.length(); i++)
			{
				char currentChar = command.charAt(i);
				if(currentChar == '[')
				{
					tracker++;
				}
				else if(currentChar == ']')
				{
					tracker--;
				}
			}
			if(tracker>0)
			{
				System.out.println("You are missing "+tracker+" ]");
			}
			if(tracker<0)
			{
				int missing = tracker/-1;
				System.out.println("You are missing "+missing+" [");
			}
			int i = 0;
			while(i<command.length())
			{
				char currentChar = command.charAt(i);
				switch(currentChar)
				{
					case '>':
						pointer++;
						if(pointer == 30000)
						{
							pointer = 0;
						}
						break;
					case '<':
						pointer--;
						if(pointer<0)
						{
							pointer = 30000;
						}
						break;
					case '+':
						tape[pointer]++;
						break;
					case '-':
						tape[pointer]--;
						break;
					case '.':
						System.out.println("OUT>" + tape[pointer]);
						break;
					case ',':
						System.out.print("BYTE>");
						tape[pointer] = Integer.parseInt(input.nextLine());
						break;
					case '[':
						break;
					case ']':
						break;
				}
				i++;
			}
			
		}

	}

}
